File: 2ndSemester/kursprogramowania/lista4/zad2/PascalsTriangleRow.java
import java.util.ArrayList;

/**
 *@class PascalsTriangleRow 
 * 
 *@brief class containing methods to create nth row of Pascal's triangle
 * 
 */
public final class PascalsTriangleRow 
{
    private ArrayList<Long> row;

    /**
     * @brief returns full nth row of Pascal's Triangle
     * 
     * @param n - should be nonnegative and smaller than 62
     * @return std::vector<int> of all binomials n over something
     */
    static public ArrayList<Long> pascalsRow(int n) throws IllegalArgumentException
    {
        if (n < 0)
        {
            throw new IllegalArgumentException("parameter should be nonnegative, got " + n);
        }
        
        if (n > 61)
        {
            throw new IllegalArgumentException("parameter should be smaller than 62, got " + n);
        }
        
        ArrayList<Long> nthRow = new ArrayList<Long>();
        nthRow.add(1l);
        
        for (int i = 0; i < n; i++)
        {
            nthRow.add(nthRow.get(i) * (long)(n - i) / (long)(i + 1));
        }
        
        return nthRow;
    }

    public PascalsTriangleRow(final int n) throws IllegalArgumentException
    {
        if (n < 0)
        {
            throw new IllegalArgumentException("parameter n should be nonnegative, got " + n);
        }
        
        if (n > 61)
        {
            throw new IllegalArgumentException("parameter n should be smaller than 62, got " + n);
        }

        row = pascalsRow(n);
    }

    public Long mthElement(final int m) throws IllegalArgumentException
    {
        if (m < 0)
        {
            throw new IllegalArgumentException("parametr m should be nonnegative, got " + m);
        }

        int n = (int) row.size() - 1;

        if (m > n) 
        {
            throw new IllegalArgumentException("parametr m should be smaller or equal to n (" + n + "), got " + m);
        }
        return row.get(m);
    }
}

File: 2ndSemester/kursprogramowania/lista4/zad3/cpp/src/pascalsTriangleRow.cpp
#include <pascalsTriangleRow.hpp>

#include <stdexcept>

std::vector<int> PascalsTriangleRow::pascalsRow(const int n) noexcept(false)
{
    if (n < 0)
        throw std::invalid_argument(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "parameter n should be nonnegative, got " + std::to_string(n));

    if (n > 33)
        throw std::out_of_range(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "parameter n should be smaller than 34, got " + std::to_string(n));

    std::vector<int> nthRow{1};

    for (int i = 0; i < n; i++)
    {
        nthRow.push_back(static_cast<int>(static_cast<long>(nthRow[static_cast<size_t>(i)])*static_cast<long>(n-i)/static_cast<long>(i+1)));
    }
    return nthRow;
}

PascalsTriangleRow::PascalsTriangleRow(const int n) noexcept(false)
{
    if (n < 0)
        throw std::invalid_argument(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "parameter n should be nonnegative, got " + std::to_string(n));

    if (n > 33)
        throw std::out_of_range(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "parameter n should be smaller than 34, got " + std::to_string(n));

    row = PascalsTriangleRow::pascalsRow(n);
}

int PascalsTriangleRow::mthElement(const int m) noexcept(false)
{
    if (m < 0)
        throw std::invalid_argument(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "parameter m should be nonnegative, got " + std::to_string(m));

    int n = static_cast<int>(this->row.size()) - 1;

    if (m > n)
        throw std::out_of_range(std::string(__FILE__) + ":" + std::to_string(__LINE__) + " [" + __func__ + "] " + "parameter m should be smaller or equal to n (" + std::to_string(n) + "), got " + std::to_string(m));
        
    return this->row[static_cast<size_t>(m)];
}
